gemrb/core/System/String.cpp

```diff
@@ -100,6 +105,49 @@ static String* StringFromEncodedData(const ieByte* string, const EncodingStruct&
 	return dbString;
 }
 
+#if HAVE_ICONV
+// returns new string converted to given encoding
+// in case iconv is not available, or there is encoding error, copy of original
+// string is returned
+char* ConvertCharEncoding(const char* string, const char* from, const char* to) {
+	iconv_t cd = iconv_open(to, from);
+	if (cd == (iconv_t)-1) {
+	    Log(ERROR, "String", "iconv_open(%s, %s) failed with error: %s", to, from, strerror(errno));
+	    return strdup(string);
+	}
+
+
+	char * in = (char *) string;
+	size_t in_len = strlen(string);
+	// twice as much bytes should be safe for most encodings
+	size_t out_len = in_len * 2 + 1;
+	size_t out_len_left = out_len;
+	char* buf = (char*) malloc(out_len);
+	char* buf_out = buf;
+	#if ICONV_ACCEPTS_NONCONST_INPUT
+		size_t ret = iconv(cd, &in, &in_len, &buf_out, &out_len_left);
+	#else
+		size_t ret = iconv(cd, (const char **)&in, &in_len, &buf_out, &out_len_left);
```

Comment:
Compiler error:
```gcc
./gemrb/gemrb/core/System/String.cpp:130:26: error: invalid conversion from 'const char**' to 'char**' [-fpermissive]
  130 |   size_t ret = iconv(cd, (const char **)&in, &in_len, &buf_out, &out_len_left);
      |                          ^~~~~~~~~~~~~~~~~~
      |                          |
      |                          const char**
In file included from ./gemrb/gemrb/core/System/String.cpp:36:
./mingw64/include/iconv.h:82:43: note:   initializing argument 2 of 'size_t libiconv(libiconv_t, char**, size_t*, char**, size_t*)'
   82 | extern size_t iconv (iconv_t cd,  char* * inbuf, size_t *inbytesleft, char* * outbuf, size_t *outbytesleft);
```

Reply:
That's odd, how come FindICONV.cmake didn't define ICONV_ACCEPTS_NONCONST_INPUT for you?

Reply:
Apparently, Windows version of `iconv` expects non-const input, and cmake didn't detect that. What your cmake log says, can you see iconv-related warnings about cmake not being able to determine `iconv` signature?

Reply:
I'll check that again, it looks as if I have some artifacts of the attempt to reproduce that SDL1-iconv-AppVeyor in place.

Reply:
Clean CMake fixed _that_.

Now my linker complains about them with `lib` prepended to these functions, like:
```gcc
./mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/Devel/Sources/Cpp/gemrb/gemrb/core/System/String.cpp:128: undefined reference to `libiconv'
```
I'll try to find out what this is about.

Reply:
I don't get this `#if` a `char**` (or any pointer for that matter) can always be passed to a function that takes the `const` variant. What am I missing?

Reply:
The const variant isn't guaranteed to be there. There's also ICONV_ACCEPTS_CONST_INPUT if we wanted to invert the check.

Reply:
I know the const variant isn't guaranteed to exist. However, `oldchar` is already a `char*` so it could be passed to either a `const` or non `const` function. I don't understand the need for this still :)

Reply:
Neither is guaranteed to exist, not just the one.

Reply:
I'm still lost. Those parameters should be good to pass to either variety... it shouldn't matter which function exists.

Reply:
Ahh, I see what you meant now!

Reply:
It was adaptation of code used in TTFImporter. @bradallred makes a valid point, I will change that.